dbms/src/Storages/StorageReplicatedMergeTree.cpp

```diff
@@ -5106,19 +5106,20 @@ void StorageReplicatedMergeTree::movePartitionToTable(const StoragePtr & dest_ta
 
         Transaction transaction(*dest_table_storage);
         {
-            auto data_parts_lock = lockParts();
+            auto dest_data_parts_lock = dest_table_storage->lockParts();
 
             for (MutableDataPartPtr & part : dst_parts)
-                dest_table_storage->renameTempPartAndReplace(part, nullptr, &transaction, data_parts_lock);
+                dest_table_storage->renameTempPartAndReplace(part, nullptr, &transaction, dest_data_parts_lock);
         }
 
         op_results = zookeeper->multi(ops);
 
         {
-            auto data_parts_lock = lockParts();
-
-            transaction.commit(&data_parts_lock);
-            parts_to_remove = removePartsInRangeFromWorkingSet(drop_range, true, false, data_parts_lock);
+            auto src_data_parts_lock = lockParts();
+            auto dest_data_parts_lock = dest_table_storage->lockParts();
+            
+            transaction.commit(&src_data_parts_lock);
```

Comment:
It is critical, that `MergeTreeData::Transaction::commit(...)` gets called under parts lock that protects the correct data parts!
But `src_data_parts_lock` is not consistent with `dest_table_storage` used in c-tor! This is incorrect.

As for the code of `MergeTreeData::Transaction::commit(...)` method.
Basically it first checks if a lock object has been passed (but not whether it is actually taken - sic!!!) and if not, it creates its own lock and proceeds with processing its precommitted parts. And here (if no lock has been passed) it becomes unclear what if there is a concurrent `MergeTreeData::Transaction`

Reply:
@Akazz I applied some changes, let me know if the changes are ok now :)

@nikitamikhaylov thank you very much for your previous fix ! 🙏